### PR TITLE
lifecycle class

### LINKED ISSUE #767: Lifecycle class
Currently the lifecycle for TM:PE is completely obfuscated across multiple classes, it's impossible to see at a glance what is going on through the various stages, and it's getting even more confusing now that we're supporting hot reloads, etc.

In #764 I suggested briefly this: https://github.com/CitiesSkylinesMods/TMPE/pull/764#issuecomment-593165213

> I think at some point we should consider creating a `Lifecycle.cs` that contains some sort of state engine for the mod lifecycle management (enabed -> loading -> unloading -> disabled) but that's something for a later date.

Kian then added this: https://github.com/CitiesSkylinesMods/TMPE/pull/764#issuecomment-593166163

> I have thought of that but I can't see how it would work. 
>  
> the only thing I can think of is:
> 1. move all code from `OnEnabled/OnDisabled`, `LoadingExtension.cs`, and `SerializeDataExtension.cs` to `lifecycle.cs` 
> 2. then all `OnEnabled/OnDisabled`, `LoadingExtension.cs`, and `SerializeDataExtension.cs`  do is to call appropriate functions of lifecycle.cs at appropriate times. 
>  
> So lifeCycle.cs acts as a wrapper for `IUserMod`, `LoadingExtension`, and `SerializableDataExtension` .  A bit ugly if you ask me.
>  
> The best thing to do is a mod for hot reload to patch CS so that it does not do weird stuff.

Even if we have mod to patch CS to not do weird stuff, we should still have a more coherent way of managing TMPE lifecycle.

Building on Kian's ideas, I suggest:

1. Lifecycle class gets stub methods for the various states
2. `TrafficManagerMod` and `LoadingExtension` (and anything else relevant) simply calls the applicable lifecycle methods
3. Rather than filling lifecycle methods with bloated code, we should atomise funcitonality in to methods in some other class
4. The lifecycle would then just call those other methods

The result would be a very concise lifecycle class and you could just read it to see exactly what steps happen in each stage of the lifecycle without getting bogged down in reams of code.

For example, enabled/disabled stuff in **`Lifecycle.cs`** could look something like (rough mockup):

```cs
// comment would note all the ways enabled can be triggered, such as:
// * mod autoenabler mod when user subscribes TMPE
// * enabled via content manager > mods
// * already enabled when game starts
// * hot-reload (can happen in-game, main menu, etc)
public void OnEnabled() {
    if (!HotReload) {
        LogEnvironmentDetails();
        PerformCompatibiltyChecks();
    }
}

// comment...
public void OnDisabled() {
    RemoveEventListeners();
    if (HotReload) {
        Patches.Remove();
        ModUI.Remove();
        //...
    }
}
```

Basically, we should be able to see at a glance wtf is going on without having to wade through gargantuan amounts of code.

The **`TrafficManagerMod.cs`** would then be like (rough mockup):

```cs
public void OnEnabled() {
    Lifecycle.OnEnabled();
}

public void OnDisabled() {
    Lifecycle.OnDisabled();
}
```

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #767
- Introduce a lifecycle class (`TMPELifecyckle`)
- the class is a singleton that is created when the TMPE mod is activated and destroyed when the TMPE mod is disabled
- moved all the static variables in lifecycle code as instant variables to the new class.
- The `TMPELifecycle`'s finalized is called when TMPE is disabled which proves there is no memory leak :) - I think if anyone uses `Singleton<TMPELifecyckle>.instance` that would create an extra reference to `TMPELifecyckle` which prevents the class from finalization leading to a memory leak.
- moved all the lifecycle related code to the lifecycle namespace (might break NC a bit)

 https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=life-cycle

~~depends on PR #1076~~

### CODE REVIEW

**Comment by kianzarrin** [comment]
I don't know why its like this:
image
I didn't move LoadingExtension.cs to TMPELifeCycle.cs

**Review by kvakvs — APPROVED** [review]
I like this

**Review by chameleon-tbn — APPROVED** [review]
lgtm

File: TLM/TLM/Lifecycle/TrafficManagerMod.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        public void OnEnabled() => TMPELifecycle.StartMod();
+
+        [UsedImplicitly]
+        public void OnDisabled() => TMPELifecycle.EndMod();
+
+        [UsedImplicitly]
+        public void OnSettingsUI(UIHelper helper) {
+            // Note: This bugs out if done in OnEnabled(), hence doing it here instead.
+            LocaleManager.eventLocaleChanged -= Translation.HandleGameLocaleChange;
+            LocaleManager.eventLocaleChanged += Translation.HandleGameLocaleChange;
```
Comment (krzychu124) [inline TLM/TLM/Lifecycle/TrafficManagerMod.cs:42]: Does not make sense...
Also OnSettingsUI is called multiple times, so I'm not sure if it's a good place to register callback

File: TLM/TLM/Lifecycle/TMPELifecycle.cs [hunk 2]
```diff
@@ -36,18 +67,13 @@ public class LoadingExtension : LoadingExtensionBase {
...
-        public static bool IsGameLoaded { get; private set; }
 
-        private static void RegisterCustomManagers() {
-            RegisteredManagers = new List<ICustomManager>();
+        private static void CheckForIncompatibleMods() {
+            ModsCompatibilityChecker mcc = new ModsCompatibilityChecker();
+            mcc.PerformModCheck();
+        }
 
+        private void RegisterCustomManagers() {
```
Comment (krzychu124) [inline TLM/TLM/Lifecycle/TMPELifecycle.cs:76]: [Bug introduced]
List was "reset" (instantiated) before change. (first call is `TMPELifecycle.Awake()` then second one in the `TMPELifecycle.Load()`)
Ended up on multiple instances in the list -> multiple calls to `OnLevelLoading()` and `OnLevelUnloading()`

**Review by krzychu124 — CHANGES_REQUESTED** [review]
Two issues in code

File: TLM/TLM/Lifecycle/TrafficManagerMod.cs (hunk 1 above)
Comment (kianzarrin) [reply TLM/TLM/Lifecycle/TrafficManagerMod.cs:42]: what does not make sense? the -= += ?

File: TLM/TLM/Lifecycle/TMPELifecycle.cs (hunk 2 above)
Comment (kianzarrin) [reply TLM/TLM/Lifecycle/TMPELifecycle.cs:76]: thanks :)

File: TLM/TLM/Lifecycle/TrafficManagerMod.cs (hunk 1 above)
Comment (kianzarrin) [reply TLM/TLM/Lifecycle/TrafficManagerMod.cs:42]: > Also OnSettingsUI is called multiple times
that's where the  -= += coms in.

**Review by DaEgi01 — APPROVED** [review]
lgtm besides the points that krzy mentioned.
I m not sure how much is won by the change, since you move global state out of an object that is instantiated by the game and used for lifecycle purposes to global state that lazily instantiates itself and is also meant for lifecycle purposes.
but currently I have no opinion on how to do it better.
the way I normally do that for CSL is also messed up bad.

File: TLM/TLM/Lifecycle/TrafficManagerMod.cs (hunk 1 above)
Comment (krzychu124) [reply TLM/TLM/Lifecycle/TrafficManagerMod.cs:42]: Ok, I misread the order. I also checked if not forget to unsubscribe. Looks good!

**Review by krzychu124 — APPROVED** [review]
Looks good! 👍

**Comment by kianzarrin** [comment]
Moved all lifecycle code to a TMPELifeCycle.

**Pull request merged**